yarn-project/archiver/src/archiver/eth_log_handlers.ts

```diff
@@ -63,61 +64,116 @@ export function processCancelledL1ToL2MessagesLogs(
+}
+
+export async function processBlockBodyLogs(
+  publicClient: PublicClient,
+  logs: Log<bigint, number, false, undefined, true, typeof AvailabilityOracleAbi, 'TxsPublished'>[],
+): Promise<[Body, Buffer][]> {
+  const retrievedBlockBodies: [Body, Buffer][] = [];
+  for (const log of logs) {
+    const newBlockBody = await getBlockBodiesFromAvailabilityOracleTx(publicClient, log.transactionHash!);
+    retrievedBlockBodies.push([newBlockBody, Buffer.from(hexToBytes(log.args.txsHash))]);
+  }
+
+  return retrievedBlockBodies;
 }
 
 /**
- * Builds an L2 block out of calldata from the tx that published it.
+ * Gets block metadata (header and archive root) from the calldata of a L1 transaction
  * Assumes that the block was published from an EOA.
  * TODO: Add retries and error management.
  * @param publicClient - The viem public client to use for transaction retrieval.
  * @param txHash - Hash of the tx that published it.
  * @param l2BlockNum - L2 block number.
- * @returns An L2 block deserialized from the calldata.
+ * @returns L2 block metadata (header and archive) from the calldata, deserialized
  */
-async function getBlockFromCallData(
+async function getBlockMetadataFromRollupTx(
   publicClient: PublicClient,
   txHash: `0x${string}`,
   l2BlockNum: bigint,
-): Promise<L2Block> {
+): Promise<[Header, AppendOnlyTreeSnapshot]> {
   const { input: data } = await publicClient.getTransaction({ hash: txHash });
   const { functionName, args } = decodeFunctionData({
     abi: RollupAbi,
     data,
   });
+
   if (functionName !== 'process') {
     throw new Error(`Unexpected method called ${functionName}`);
   }
-  const [headerHex, archiveRootHex, bodyHex] = args! as [Hex, Hex, Hex, Hex];
-  const blockBuffer = Buffer.concat([
-    Buffer.from(hexToBytes(headerHex)),
-    Buffer.from(hexToBytes(archiveRootHex)), // L2Block.archive.root
-    numToUInt32BE(Number(l2BlockNum)), // L2Block.archive.nextAvailableLeafIndex
-    Buffer.from(hexToBytes(bodyHex)),
-  ]);
-  const block = L2Block.fromBuffer(blockBuffer);
-  if (BigInt(block.number) !== l2BlockNum) {
-    throw new Error(`Block number mismatch: expected ${l2BlockNum} but got ${block.number}`);
+  const [headerHex, archiveRootHex] = args! as readonly [Hex, Hex, Hex, Hex];
+
+  const header = Header.fromBuffer(Buffer.from(hexToBytes(headerHex)));
+
+  const blockNumberFromHeader = header.globalVariables.blockNumber.toBigInt();
+
+  if (blockNumberFromHeader !== l2BlockNum) {
+    throw new Error(`Block number mismatch: expected ${l2BlockNum} but got ${blockNumberFromHeader}`);
+  }
+
+  const archive = AppendOnlyTreeSnapshot.fromBuffer(
+    Buffer.concat([
+      Buffer.from(hexToBytes(archiveRootHex)), // L2Block.archive.root
+      numToUInt32BE(Number(l2BlockNum)), // L2Block.archive.nextAvailableLeafIndex
+    ]),
+  );
+
+  return [header, archive];
+}
+
+/**
+ * Gets block bodies from calldata of a L1 transaction, and deserializes them into Body objects
```

Comment:
```suggestion
 * Gets block bodies from calldata of an L1 transaction, and deserializes them into Body objects.
```

Pardon me for my autism